debug_lib.py

```diff
@@ -0,0 +1,178 @@
+
+    Related to train_lib.TrainManager
+  '''
+
+  DEFAULT_CONFIG = dict(
+      table_length=5e4,
+      table_name='debug_test',
+  )
+
+  def __init__(self, 
+      learner, 
+      data_source, 
+      saved_model_path, 
+      table_length,
+      table_name,
+    ):
+    self.learner = learner
+    self.data_source = data_source
+    self.hidden_state = learner.policy.initial_state(data_source.batch_size)
+    self.saved_model_path = saved_model_path
+
+    self.data_profiler = utils.Profiler()
+    self.step_profiler = utils.Profiler()
+
+    self.mdf = pd.DataFrame()
+    self.time = 0
+    self.table_length = table_length
+    self.table_name = table_name
+
+  def step(self):
+    with self.data_profiler:
+      batch = sanitize_batch(next(self.data_source))
+    with self.step_profiler:
+      # debug_namedtuple = self.learner.step(batch, self.hidden_state)
+      debug_namedtuple = self.learner.compiled_step(batch, self.hidden_state)
+      self.hidden_state = debug_namedtuple.final_state
+    self.update_table(debug_namedtuple)
+    return debug_namedtuple.loss
+
+  def update_table(self, debug_namedtuple):
+    nests = {
+      'obs': debug_namedtuple.observations, 
+      'sample': debug_namedtuple.samples,
+      'gamestate': debug_namedtuple.gamestate,
+      'dist': debug_namedtuple.distances
+    }
+    path_delim = '.'
+
+    df = None
+    for nest_name, nest in nests.items():
+
+      # E.g., skip dist for controller_heads that do not yet provide controller component-wise distance
+      if nest_name == 'dist' and not tf.nest.is_nested(nest):
+        continue
+
+      for path, array in tree.flatten_with_path(nest):
+        leaf_name = path_delim.join([str(subpath) for subpath in path])
+
+        wide_single_df = pd.DataFrame(tf.squeeze(array).numpy().astype(float))
+        wide_single_df['Time'] = wide_single_df.index
+        col_nm = f'{nest_name} {leaf_name}'
+        long_df = pd.melt(wide_single_df, id_vars=['Time'],
+          var_name='Batch', value_name=col_nm)
+
+        if df is not None:
+          df[col_nm] = long_df[col_nm]
+        else:
+          df = long_df
+
+    df['Time'] += self.time
+    self.time += len(wide_single_df)
+    df['Loss'] = df[[col for col in df.columns if 'dist ' in col]].agg(np.nansum, axis='columns')
+
+    self.mdf = self.mdf.append(df)
+
+    if len(self.mdf) > self.table_length:
+      print(f'Reached or exceeded specified debug.max_table_size: {self.table_length}')
+      out_fn = self.saved_model_path + f'{self.table_name}.csv'
+      self.mdf.to_csv(out_fn)
+      print(f'Wrote {len(self.mdf)} frames to {out_fn}')
+      print(f'Exiting.')
+      sys.exit(0)
+    return
+
+
+class DebugLearner:
+  '''
+    Related to learner.Learner
+  '''
+
+  DEFAULT_CONFIG = dict(
+  )
+
+  def __init__(self,
+      policy: Policy):
+    self.policy = policy
+    self.compiled_step = tf.function(self.step)
+    self.sample = lambda *structs: policy.sample(*tf.nest.flatten(structs))
```

Comment:
policy.sample doesn't work for me here. I'm using the same sample function defined in eval.py

Reply:
eval.py needs to do this because it is loading from a saved model, which for silly reasons needs this as a workaround.

What is the error you're seeing? You can remove `@tf.funtion`s to make debugging easier.

Reply:
The error is `ValueError: Structure of Python function inputs does not match input_signature`, with input being an ordered dict nest and expected signature being a flattened nest. The debug scripts load a saved model and run it on the test set, recording model inputs, outputs, etc., so it should need the same code as `eval.py` for sampling.

Reply:
Ah ok, I missed that you were loading a saved model. My b.